### ISSUE TITLE
Low CPU efficiency for output.nproc=128 on Perlmutter

### ISSUE THREAD

**Issue by jchiang87** [LSSTDESC/imSim#324]
Now that the atmospheric screens can be shared between processes, the memory per process is less than 4GB, and we can run avoiding idle cores on the Perlmutter CPU nodes and have 128 CCDs from the same visit being processed concurrently.  Unfortunately, when I tried this, the CPU usage was much less than 100% on average.  Here's a screen shot of `top` on a Perlmutter node with 128 CCDs running at once:
<img src="https://user-images.githubusercontent.com/1994473/218336442-98df4cf5-1c7d-4681-aed3-d6fe10bb868a.png" height="500">
The overall efficiency and resulting throughput was substantially less than running ~70 CCDs as separate galsim processes, each of which generally runs near 100%.

**Comment by cwwalter** [comment]
I wonder if the fact that there is one of them with high usage means that it is the dispatching process and it is blocking some work?

**Comment by jmeyers314** [comment]
Are you running with `OMP_NUM_THREADS=1` in the environment?  If not, then `batoid` is likely spawning 128 threads per process (16384 per node), which I'm sure will slow things down a bunch.

**Comment by rmjarvis** [comment]
Where is that being done?  We can wrap batoid calls in a [single_threaded](http://galsim-developers.github.io/GalSim/_build/html/misc.html?highlight=single#galsim.utilities.single_threaded) context.  That's probably better than relying on users setting environment variables appropriately.

**Comment by jchiang87** [comment]
> Are you running with OMP_NUM_THREADS=1 in the environment?

yes.

**Comment by jmeyers314** [comment]
Just noticed that even with OMP_NUM_THREADS=1 set for the current command or exported to the env, imSim with nproc=8 will completely saturate 128 cores at USDF (I'm guessing this is batoid but didn't check).  Maybe OMP_NUM_THREADS doesn't penetrate to subprocesses?

We can try the single_threaded context Mike suggested, but I think I've seen enough now to go and at least limit the default batoid threading to something like 4 or 8 (and maybe even just set it to 1).

**Comment by cwwalter** [comment]
Did you see the slack discussion about OMP_PROC_BIND="false"?

https://lsstc.slack.com/archives/CHXKSF3HC/p1682381649221819

**Comment by jmeyers314** [comment]
> Just noticed that even with OMP_NUM_THREADS=1 ...

Ignore this.  Just noticed in my command history that I typoed `OPM_NUM_THREADS=1`.  Spelling that correctly looks like it works.

**Comment by cwwalter** [comment]
Upping this (and see related discussion yesterday in slack).  Both Jim and I are seeing this now (for me at least on USDF).

After the affinity bug fix, my 105 core job for 1/2 a focal plane started up with ~99% utilization per node.

Over time this started to degrade.  Far into the job it had dropped down to more like 80% and the node became completely unresponsive / unusable.  (on USDF you can ssh into batch nodes where you have jobs running).   You could barely type / run top etc and the batch systems were not able to really communicate with the node.

Although no process seemed to be using a lot of memory the overall memory usage on the node was very high and the load average was about 5X when it was in the early parts of the job.

I seems to me a memory leak in the C++ portions of the code is a plausible explanation for this.  Galsim has been used a lot in DES etc, so it is probably something specific to what we have added.  Two thing that come to mind are the recent changes in our silicon model in GalSim and also the C++ portion of batoid.

**Comment by jchiang87** [comment]
Running a small two process job on a shared node at USDF, I got the following error:
```
slurmstepd: error: Detected 6 oom-kill event(s) in StepId=24318542.0. Some of your processes may have been killed by the cgroup out-of-memory handler.
srun: error: sdfrome038: task 0: Out Of Memory
```
I think this occurred because I only asked for 4GB per core (this is the default if one doesn't specify anything and just sets `--cpus-per-task`), and at least one of the processes in my job exceeded that number.  In slurm, one can request more memory with `--mem-per-cpu` to avoid this.

Anyway, it's not clear to me that this particular oom error is related to the slow down, since that still happens when I have a node exclusively, and there's no constraint on the per process memory usage aside from the overall RAM available on the node.

**Comment by rmjarvis** [comment]
Memory leaks in C++ seem mildly unlikely to me.  (Everything in C++ is in shared_ptr's so it's actually not too easy to get a memory leak, but maybe we missed something.)  It's somewhat easier to memory leaks in python, since references can stick around in weird places that can be hard to notice.

**Comment by cwwalter** [comment]
This was fixed in #418.  The problem was that all the stamps were being accumulated in a loop when we had checkpointing turned off.  Now we always batch the results into 100 pieces so that the memory can't build up.
